config-controller/internal/controller/policy_controller.go

```diff
@@ -62,8 +63,48 @@ func (r *SecurityPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 
 	if ok, err := policyCR.Spec.IsValid(); !ok {
 		return ctrl.Result{}, errors.Wrapf(err, "Invalid policy resource: namespace=%s, name=%s", req.Namespace, req.Name)
+	}
 
+	if err := r.PolicyClient.EnsureFresh(ctx); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "Failed to refresh")
 	}
+
+	// Ensure a finalazer is added to each policy custom resource if it is not being deleted
+	if policyCR.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The policy is not being deleted, so if it does not have our finalizer,
+		// then lets update the policy to add the finalizer. This is equivalent
+		// to registering our finalizer in preparation for a future delete.
+		if !controllerutil.ContainsFinalizer(policyCR, policyFinalizer) {
+			controllerutil.AddFinalizer(policyCR, policyFinalizer)
+			if err := r.K8sClient.Update(ctx, policyCR); err != nil {
+				return ctrl.Result{}, errors.Wrapf(err, "unable to add finalizer to policy %q", policyCR.GetName())
+			}
+		}
+	} else {
+		// The policy is being deleted since k8s set the deletion timestamp because a finalizer had already been added
+		if controllerutil.ContainsFinalizer(policyCR, policyFinalizer) {
+			// finalizer is present, so lets handle the external dependency of deleting policy in central
+			if err := r.PolicyClient.DeletePolicy(ctx, policyCR.GetName()); err != nil {
+				// if we failed to delete the policy in central, return with error
+				// so that it can be retried.
+				return ctrl.Result{}, errors.Wrapf(err, "failed to delete policy %q", policyCR.GetName())
+
+			}
+
+			// remove our finalizer from the list and update it.
+			controllerutil.RemoveFinalizer(policyCR, policyFinalizer)
+			if err := r.K8sClient.Update(ctx, policyCR); err != nil {
+				return ctrl.Result{}, errors.Wrapf(err, "failed to delete policy %q", policyCR.GetName())
```

Comment:
Should we have a different error message here?  Perhaps more specific to the removal of the finalizer?